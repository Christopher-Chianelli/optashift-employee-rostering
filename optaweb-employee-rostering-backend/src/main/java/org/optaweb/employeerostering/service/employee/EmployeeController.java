/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.optaweb.employeerostering.service.employee;

import java.io.IOException;
import java.util.List;

import javax.enterprise.context.ApplicationScoped;
import javax.inject.Inject;
import javax.validation.Valid;
import javax.validation.constraints.Min;
import javax.ws.rs.Consumes;
import javax.ws.rs.DELETE;
import javax.ws.rs.GET;
import javax.ws.rs.POST;
import javax.ws.rs.PUT;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;

import org.jboss.resteasy.annotations.providers.multipart.MultipartForm;
import org.optaweb.employeerostering.domain.common.MultipartBody;
import org.optaweb.employeerostering.domain.employee.Employee;
import org.optaweb.employeerostering.domain.employee.view.EmployeeAvailabilityView;
import org.optaweb.employeerostering.domain.employee.view.EmployeeView;

@Path("/rest/tenant/{tenantId}/employee")
// @Api(tags = "Employee")
@ApplicationScoped
public class EmployeeController {

    private final EmployeeService employeeService;

    @Inject
    public EmployeeController(EmployeeService employeeService) {
        this.employeeService = employeeService;
    }

    // ************************************************************************
    // Employee
    // ************************************************************************

    // @ApiOperation("Get a list of all employees")
    @GET
    @Path("/")
    public List<Employee> getEmployeeList(@PathParam("tenantId") @Min(0) Integer tenantId) {
        return employeeService.getEmployeeList(tenantId);
    }

    // @ApiOperation("Get an employee by id")
    @GET
    @Path("/{id}")
    public Employee getEmployee(@PathParam("tenantId") @Min(0) Integer tenantId,
            @PathParam("id") @Min(0) Long id) {
        return employeeService.getEmployee(tenantId, id);
    }

    // @ApiOperation("Delete an employee")
    @DELETE
    @Path("/{id}")
    public Boolean deleteEmployee(@PathParam("tenantId") @Min(0) Integer tenantId,
            @PathParam("id") @Min(0) Long id) {
        return employeeService.deleteEmployee(tenantId, id);
    }

    // @ApiOperation("Add a new employee")
    @POST
    @Path("/add")
    public Employee createEmployee(@PathParam("tenantId") @Min(0) Integer tenantId,
            @Valid EmployeeView employeeView) {
        return employeeService.createEmployee(tenantId, employeeView);
    }

    // @ApiOperation("Import employees from an Excel file")
    @POST
    @Path("/import")
    @Consumes(MediaType.MULTIPART_FORM_DATA)
    @Produces(MediaType.APPLICATION_JSON)
    public List<Employee> addEmployeesFromExcelFile(@PathParam("tenantId") @Min(0) Integer tenantId,
            @MultipartForm MultipartBody excelDataFile)
            throws IOException {

        return employeeService.importEmployeesFromExcel(tenantId, excelDataFile.file);
    }

    // @ApiOperation("Update an employee")
    @POST
    @Path("/update")
    public Employee updateEmployee(@PathParam("tenantId") @Min(0) Integer tenantId,
            @Valid EmployeeView employeeView) {
        return employeeService.updateEmployee(tenantId, employeeView);
    }

    // ************************************************************************
    // EmployeeAvailability
    // ************************************************************************

    // @ApiOperation("Get an employee availability by id")
    @GET
    @Path("/availability/{id}")
    public EmployeeAvailabilityView getEmployeeAvailability(@PathParam("tenantId") @Min(0) Integer tenantId,
            @PathParam("id") @Min(0) Long id) {
        return employeeService.getEmployeeAvailability(tenantId, id);
    }

    // @ApiOperation("Add a new employee availability")
    @POST
    @Path("/availability/add")
    public EmployeeAvailabilityView createEmployeeAvailability(@PathParam("tenantId") @Min(0) Integer tenantId,
            @Valid EmployeeAvailabilityView employeeAvailabilityView) {
        return employeeService.createEmployeeAvailability(tenantId, employeeAvailabilityView);
    }

    // @ApiOperation("Update an employee availability")
    @PUT
    @Path("/availability/update")
    public EmployeeAvailabilityView updateEmployeeAvailability(@PathParam("tenantId") @Min(0) Integer tenantId,
            @Valid EmployeeAvailabilityView employeeAvailabilityView) {
        return employeeService.updateEmployeeAvailability(tenantId, employeeAvailabilityView);
    }

    // @ApiOperation("Delete an employee availability")
    @DELETE
    @Path("/availability/{id}")
    public Boolean deleteEmployeeAvailability(@PathParam("tenantId") @Min(0) Integer tenantId,
            @PathParam("id") @Min(0) Long id) {
        return employeeService.deleteEmployeeAvailability(tenantId, id);
    }
}
