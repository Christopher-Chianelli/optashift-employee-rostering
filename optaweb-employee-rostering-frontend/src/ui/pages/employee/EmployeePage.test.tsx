/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import { Employee } from 'domain/Employee';
import { getRouterProps } from 'util/BookmarkableTestUtils';
import { FileUpload } from '@patternfly/react-core';
import { Map } from 'immutable';
import { mockStore } from 'store/mockStore';
import { Contract } from 'domain/Contract';
import DomainObjectView from 'domain/DomainObjectView';
import { mockRedux } from 'setupTests';
import { employeeSelectors } from 'store/employee';
import { EditableEmployeeRow, EmployeeRow, EmployeesPage } from './EmployeesPage';

const noEmployeesNoContractsStore = mockStore({
  skillList: {
    isLoading: false,
    skillMapById: Map(),
  },
  contractList: {
    isLoading: false,
    contractMapById: Map(),
  },
  employeeList: {
    isLoading: false,
    employeeMapById: Map(),
  },
}).store;

const contract: Contract = {
  id: 0,
  tenantId: 0,
  name: 'Contract',
  maximumMinutesPerDay: null,
  maximumMinutesPerWeek: null,
  maximumMinutesPerMonth: null,
  maximumMinutesPerYear: null,
};

const noEmployeesOneContractsStore = mockStore({
  skillList: {
    isLoading: false,
    skillMapById: Map(),
  },
  contractList: {
    isLoading: false,
    contractMapById: Map<number, Contract>()
      .set(0, contract),
  },
  employeeList: {
    isLoading: false,
    employeeMapById: Map(),
  },
}).store;

const twoEmployeesOneContractsStore = mockStore({
  skillList: {
    isLoading: false,
    skillMapById: Map(),
  },
  contractList: {
    isLoading: false,
    contractMapById: Map<number, Contract>()
      .set(0, contract),
  },
  employeeList: {
    isLoading: false,
    employeeMapById: Map<number, DomainObjectView<Employee>>()
      .set(1, {
        tenantId: 0,
        id: 1,
        name: 'Employee 1',
        contract: contract.id as number,
        shortId: 'E1',
        color: '#FF0000',
        skillProficiencySet: [],
      }).set(2, {
        tenantId: 0,
        id: 2,
        name: 'Employee 2',
        contract: contract.id as number,
        shortId: 'E2',
        color: '#00FF00',
        skillProficiencySet: [],
      }),
  },
}).store;


describe('Employees page', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly with no employees', () => {
    mockRedux(noEmployeesOneContractsStore);
    const employeesPage = shallow(<EmployeesPage {...getRouterProps('/0/employee', {})} />);
    expect(toJson(employeesPage)).toMatchSnapshot();
  });

  it('should render correctly with a few employees', () => {
    mockRedux(twoEmployeesOneContractsStore);
    const employeesPage = shallow(<EmployeesPage {...getRouterProps('/0/employee', {})} />);
    expect(toJson(employeesPage)).toMatchSnapshot();
  });

  it('should render the viewer correctly', () => {
    const employee = employeeSelectors.getEmployeeById(twoEmployeesOneContractsStore.getState(), 2);
    mockRedux(twoEmployeesOneContractsStore);
    getRouterProps('/0/employees', {});
    const viewer = shallow(<EmployeeRow {...employee} />);
    expect(toJson(viewer)).toMatchSnapshot();
  });

  it('should render the editor correctly', () => {
    const employee = employeeSelectors.getEmployeeById(twoEmployeesOneContractsStore.getState(), 2);
    mockRedux(twoEmployeesOneContractsStore);
    getRouterProps('/0/employees', {});
    const editor = shallow(<EditableEmployeeRow employee={employee} isNew={false} onClose={jest.fn()} />);
    expect(toJson(editor)).toMatchSnapshot();
  });

  it('should upload the file on Excel input', () => {
    mockRedux(noEmployeesNoContractsStore);
    const employeesPage = shallow(<EmployeesPage {...getRouterProps('/0/employee', {})} />);
    const file = new File([], 'hello.xlsx');
    employeesPage.find(FileUpload).simulate('change', file);
    // expect(noEmployees.uploadEmployeeList).toBeCalledWith(file);
  });

  it('should show an error on non-excel input', () => {
    mockRedux(noEmployeesNoContractsStore);
    const employeesPage = shallow(<EmployeesPage {...getRouterProps('/0/employee', {})} />);
    employeesPage.find(FileUpload).simulate('change', '');
    // expect(noEmployees.uploadEmployeeList).not.toBeCalled();
    // expect(noEmployees.showErrorMessage).toBeCalledWith('badFileType', { fileTypes: 'Excel (.xlsx)' });
  });
});
