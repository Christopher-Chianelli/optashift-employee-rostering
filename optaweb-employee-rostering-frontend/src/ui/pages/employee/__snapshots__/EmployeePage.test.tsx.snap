// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Employees page should render correctly with a few employees 1`] = `
<Fragment>
  <div>
    <FileUpload
      dropzoneProps={
        Object {
          "accept": ".xlsx",
        }
      }
      id="file"
      name="file"
      onChange={[Function]}
    />
  </div>
  <TheTable
    columns={
      Array [
        Object {
          "name": "Trans(i18nKey=name)",
          "sorter": [Function],
        },
        Object {
          "name": "Trans(i18nKey=contract)",
          "sorter": [Function],
        },
        Object {
          "name": "Trans(i18nKey=skillProficiencies)",
        },
        Object {
          "name": "Trans(i18nKey=shortId)",
          "sorter": [Function],
        },
        Object {
          "name": "Trans(i18nKey=color)",
        },
      ]
    }
    filterText=""
    filteredRows={
      Immutable.List [
        Object {
          "color": "#FF0000",
          "contract": Object {
            "id": 0,
            "maximumMinutesPerDay": null,
            "maximumMinutesPerMonth": null,
            "maximumMinutesPerWeek": null,
            "maximumMinutesPerYear": null,
            "name": "Contract",
            "tenantId": 0,
          },
          "id": 1,
          "name": "Employee 1",
          "shortId": "E1",
          "skillProficiencySet": Array [],
          "tenantId": 0,
        },
        Object {
          "color": "#00FF00",
          "contract": Object {
            "id": 0,
            "maximumMinutesPerDay": null,
            "maximumMinutesPerMonth": null,
            "maximumMinutesPerWeek": null,
            "maximumMinutesPerYear": null,
            "name": "Contract",
            "tenantId": 0,
          },
          "id": 2,
          "name": "Employee 2",
          "shortId": "E2",
          "skillProficiencySet": Array [],
          "tenantId": 0,
        },
      ]
    }
    history={
      Object {
        "action": "PUSH",
        "block": [MockFunction],
        "createHref": [MockFunction],
        "go": [MockFunction],
        "goBack": [MockFunction],
        "goForward": [MockFunction],
        "length": 1,
        "listen": [MockFunction],
        "location": Object {
          "hash": "",
          "pathname": "/0/employee",
          "search": "",
          "state": undefined,
        },
        "push": [MockFunction],
        "replace": [MockFunction],
      }
    }
    isReversed={false}
    itemsPerPage={10}
    location={
      Object {
        "hash": "",
        "pathname": "/0/employee",
        "search": "",
        "state": undefined,
      }
    }
    match={
      Object {
        "isExact": true,
        "params": Object {},
        "path": "/0/employee",
        "url": "localhost:8080/0/employee",
      }
    }
    newRowWrapper={[Function]}
    numOfFilteredRows={2}
    onSorterChange={[Function]}
    page={1}
    rowWrapper={[Function]}
    rowsInPage={
      Immutable.List [
        Object {
          "color": "#FF0000",
          "contract": Object {
            "id": 0,
            "maximumMinutesPerDay": null,
            "maximumMinutesPerMonth": null,
            "maximumMinutesPerWeek": null,
            "maximumMinutesPerYear": null,
            "name": "Contract",
            "tenantId": 0,
          },
          "id": 1,
          "name": "Employee 1",
          "shortId": "E1",
          "skillProficiencySet": Array [],
          "tenantId": 0,
        },
        Object {
          "color": "#00FF00",
          "contract": Object {
            "id": 0,
            "maximumMinutesPerDay": null,
            "maximumMinutesPerMonth": null,
            "maximumMinutesPerWeek": null,
            "maximumMinutesPerYear": null,
            "name": "Contract",
            "tenantId": 0,
          },
          "id": 2,
          "name": "Employee 2",
          "shortId": "E2",
          "skillProficiencySet": Array [],
          "tenantId": 0,
        },
      ]
    }
    sortByIndex={0}
    title="Trans(i18nKey=employees)"
  />
</Fragment>
`;

exports[`Employees page should render correctly with no employees 1`] = `
<Fragment>
  <div>
    <FileUpload
      dropzoneProps={
        Object {
          "accept": ".xlsx",
        }
      }
      id="file"
      name="file"
      onChange={[Function]}
    />
  </div>
  <TheTable
    columns={
      Array [
        Object {
          "name": "Trans(i18nKey=name)",
          "sorter": [Function],
        },
        Object {
          "name": "Trans(i18nKey=contract)",
          "sorter": [Function],
        },
        Object {
          "name": "Trans(i18nKey=skillProficiencies)",
        },
        Object {
          "name": "Trans(i18nKey=shortId)",
          "sorter": [Function],
        },
        Object {
          "name": "Trans(i18nKey=color)",
        },
      ]
    }
    filterText=""
    filteredRows={Immutable.List []}
    history={
      Object {
        "action": "PUSH",
        "block": [MockFunction],
        "createHref": [MockFunction],
        "go": [MockFunction],
        "goBack": [MockFunction],
        "goForward": [MockFunction],
        "length": 1,
        "listen": [MockFunction],
        "location": Object {
          "hash": "",
          "pathname": "/0/employee",
          "search": "",
          "state": undefined,
        },
        "push": [MockFunction],
        "replace": [MockFunction],
      }
    }
    isReversed={false}
    itemsPerPage={10}
    location={
      Object {
        "hash": "",
        "pathname": "/0/employee",
        "search": "",
        "state": undefined,
      }
    }
    match={
      Object {
        "isExact": true,
        "params": Object {},
        "path": "/0/employee",
        "url": "localhost:8080/0/employee",
      }
    }
    newRowWrapper={[Function]}
    numOfFilteredRows={0}
    onSorterChange={[Function]}
    page={1}
    rowWrapper={[Function]}
    rowsInPage={Immutable.List []}
    sortByIndex={0}
    title="Trans(i18nKey=employees)"
  />
</Fragment>
`;

exports[`Employees page should render the editor correctly 1`] = `
<TableRow>
  <TableCell
    columnName="Trans(i18nKey=name)"
  >
    <TextInput
      onChange={[Function]}
      value="Employee 2"
    />
  </TableCell>
  <TableCell
    columnName="Trans(i18nKey=contract)"
  >
    <TypeaheadSelectInput
      emptyText="Trans(i18nKey=selectAContract)"
      noClearButton={true}
      onChange={[Function]}
      optionToStringMap={[Function]}
      options={
        Array [
          Object {
            "id": 0,
            "maximumMinutesPerDay": null,
            "maximumMinutesPerMonth": null,
            "maximumMinutesPerWeek": null,
            "maximumMinutesPerYear": null,
            "name": "Contract",
            "tenantId": 0,
          },
        ]
      }
      value={
        Object {
          "id": 0,
          "maximumMinutesPerDay": null,
          "maximumMinutesPerMonth": null,
          "maximumMinutesPerWeek": null,
          "maximumMinutesPerYear": null,
          "name": "Contract",
          "tenantId": 0,
        }
      }
    />
  </TableCell>
  <TableCell
    columnName="Trans(i18nKey=skillProficiencies)"
  >
    <MultiTypeaheadSelectInput
      emptyText="Trans(i18nKey=selectSkillProficiencies)"
      onChange={[Function]}
      optionToStringMap={[Function]}
      options={Array []}
      value={Array []}
    />
  </TableCell>
  <TableCell
    columnName="Trans(i18nKey=shortId)"
  >
    <TextInput
      onChange={[Function]}
      value="E2"
    />
  </TableCell>
  <TableCell
    columnName="Trans(i18nKey=color)"
  >
    <ColorPicker
      currentColor="#00FF00"
      onChangeColor={[Function]}
    />
  </TableCell>
  <RowEditButtons
    isValid={true}
    onClose={[Function]}
    onSave={[Function]}
  />
</TableRow>
`;

exports[`Employees page should render the viewer correctly 1`] = `
<TableRow>
  <TableCell
    columnName="Trans(i18nKey=name)"
  >
    <Flex>
      <FlexItem>
        <Text>
          Employee 2
        </Text>
      </FlexItem>
      <FlexItem>
        <Button
          onClick={[Function]}
          variant="link"
        >
          <ArrowIcon
            color="currentColor"
            noVerticalAlign={false}
            size="sm"
          />
        </Button>
      </FlexItem>
    </Flex>
  </TableCell>
  <TableCell
    columnName="Trans(i18nKey=contract)"
  >
    <Text>
      Contract
    </Text>
  </TableCell>
  <TableCell
    columnName="Trans(i18nKey=skillProficiencies)"
  >
    <ChipGroup
      aria-label="Chip group category"
      categoryName=""
      closeBtnAriaLabel="Close chip group"
      collapsedText="\${remaining} more"
      defaultIsOpen={false}
      expandedText="Show Less"
      isClosable={false}
      numChips={3}
      onClick={[Function]}
      onOverflowChipClick={[Function]}
      tooltipPosition="top"
    />
  </TableCell>
  <TableCell
    columnName="Trans(i18nKey=shortId)"
  >
    <Text>
      E2
    </Text>
  </TableCell>
  <TableCell
    columnName="Trans(i18nKey=color)"
  >
    <ColorPicker
      currentColor="#00FF00"
      isDisabled={true}
      onChangeColor={[Function]}
    />
  </TableCell>
  <RowViewButtons
    onDelete={[Function]}
    onEdit={[Function]}
  />
</TableRow>
`;
